### PR TITLE
Fix memory leak issue

### LINKED ISSUE #826: Memory leak
screenshot from 2018-04-30 21-34-30
screenshot from 2018-04-30 21-36-17

### PR DESCRIPTION
**Pull request by semuxdev**
Resolves #826 

- [x] Fix vote cached validation result
- [x] Fix connection counter
- [ ] ~~Fix log event piles up~~

### CODE REVIEW

File: src/main/java/org/semux/net/ConnectionLimitHandler.java [hunk 1]
```diff
@@ -41,7 +44,7 @@ public ConnectionLimitHandler(int maxConnectionsPerIp) {
     public void channelActive(ChannelHandlerContext ctx) throws Exception {
         super.channelActive(ctx);
         InetAddress address = ((InetSocketAddress) ctx.channel().remoteAddress()).getAddress();
-        if (connectionCount.computeIfAbsent(address, k -> new AtomicInteger(0))
+        if (connectionCount.get(address, k -> new AtomicInteger(0))
```
Comment (witoldsz) [inline src/main/java/org/semux/net/ConnectionLimitHandler.java]: When cache goes beyond `maximumSize`, the connectionCount for given address will drop down to zero.
Is that okay?

Comment (semuxdev) [reply src/main/java/org/semux/net/ConnectionLimitHandler.java]: Given that the LRU cache is sufficiently large and the keys are `InetAddress`, this should be fine. However, we may need to invalid a counter if its value goes below zero.

**Comment by semuxdev** [comment]
@witoldsz @cryptokat @orogvany any idea why the log events pile up?

**Comment by cryptokat** [comment]
@semuxdev How long does it take to reproduce the memory leak of Log4j? `Log4jLogEvent` objects are normally garbage-collected as I observed in VisualVM.

**Comment by semuxdev** [comment]
It’s very hard to reproduce. The dump we got was from one seed node after running months.

**Comment by orogvany** [comment]
Hm, why would logger hold reference to VoteSet?  Are we passing Voteset objects into logger as param and trying to rely on tostring formatting or some such?  

Or am I misunderstanding the logger memory leak?

**Comment by cryptokat** [comment]
@orogvany Check the heap dump in https://github.com/semuxproject/semux/issues/826 where ` Log4jLogEvent` objects occupies 31 megabytes of memory.

**Comment by orogvany** [comment]
I don't see the heap dump posted, just some screen shots which doesnt include the expanded Log4JLogEvent?  Can someone post or send me the heapdump?

File: src/main/java/org/semux/net/ConnectionLimitHandler.java (hunk 1 above)
Comment (witoldsz) [reply src/main/java/org/semux/net/ConnectionLimitHandler.java]: The question is: why is the LRU cache used for `connectionCount`? LRU does not seem to fit well here.

**Pull request merged**